torch/fx/experimental/Partitioner.py

```diff
@@ -0,0 +1,145 @@
+        self.children.append(partition)
+
+    def get_children(self) -> List['Partition']:
+        return self.children
+
+    def get_parents(self) -> List['Partition']:
+        return self.parents
+
+    def get_input_nodes(self) -> List[Node]:
+        """Input nodes are coming from two places:
+        placeholder and output from its parents output.
+        """
+        input_nodes: List[Node] = []
+        for node in self.nodes:
+            if node.op == 'placeholder':
+                input_nodes.append(node)
+        for partition in self.parents:
+            input_nodes += partition.get_output_nodes()
+        return input_nodes
+
+    def get_output_nodes(self) -> List[Node]:
+        """Output nodes are the nodes that without any user inside this partition."""
+        output_nodes: List[Node] = []
+        for node in self.nodes:
+            index = self.graph_module.graph.nodes.index(node)
+            user_indexes = GraphManipulation.get_all_users_of(self.graph_module, index)
+            user_nodes = set([self.graph_module.graph.nodes[i] for i in user_indexes])
+            #check if user nodes has an intersection with self.nodes
+            if not set(self.nodes).intersection(user_nodes):
+                output_nodes.append(node)
+        return output_nodes
+
+    def __str__(self) -> str:
+        if self.partition_id == 0:
+            return "root_partition"
+        line: str = 'partition id: ' + str(self.partition_id) + '\n'
+        line += 'all nodes in this partition: '
+        for node in self.nodes:
+            line += '(' + node.name + ':' + node.op + ') '
+        line += '\n'
+        line += 'parent partitions:'
+        for parent in self.parents:
+            if parent.partition_id == 0:
+                line += 'root_partition '
+            else:
+                line += 'partition_' + str(parent.partition_id) + ' '
+        line += '\n'
+        line += 'children partitions:'
+        for child in self.children:
+            line +=  str(child.partition_id) + ' '
+        line += '\n'
+        line += 'input nodes: '
+        input_nodes = self.get_input_nodes()
+        for node in input_nodes:
+            line += '(' + node.name + ':' + node.op + ') '
+        line += '\n'
+        line += 'output nodes: '
+        output_nodes = self.get_output_nodes()
+        for node in output_nodes:
+            line += '(' + node.name + ':' + node.op + ') '
+        return line
+
+class Partitioner:
+    def __init__(self) -> None:
+        """
+        After a partitioner is created,
+        it first check if multiple types of devices (backends) are involved.
+        So far, we assume that there is only one backend and one device with unlimited memory
+        """
+        self.partitions = []
+        self.devices = []
+        self.graph_module = None
+        self.node_to_partitions = {}
+
+    def partition_graph(self, fx_module: GraphModule, devices: List[str]) -> Dict[Partition, List[Partition]]:
```

Comment:
Devices as a list of string will work for now, but you'll need to add available memory at some point. We can keep it as a list of strings for now it changing it later isn't too much trouble. Also the return type here is a Partition, Is the partition meant to be an external interface data structure or is it used internally? Basically will it be a stable interface or will it need to be edited to update the function of the partitioner itself?

Reply:
The return type should be our DAG structure

Reply:
What about a list of tuples. Say [("device_0", memory_size)]

Reply:
tuples work, but I'd prefer a dict, easier to read as a user of the library.